### PR TITLE
Add fileMarks Property to Ytt Template Step

### LINKED ISSUE #121: Support for file-marks or rawOptions in App-spec for ytt
@jmcclenny-PRE commented on [Tue Mar 02 2021](https://github.com/vmware-tanzu/carvel-ytt/issues/311)

**Describe the problem/challenge you have**
I have a need to utilize file-marks within the app-spec for kapp-controller.

**Describe the solution you'd like**
Would be nice to either have file-marks as a first class field or pass in via rawOptions
eg.
```yaml
template:
  - ytt:
      ignoreUnknownComments: true
      inline:
        pathsFrom:
        - secretRef:
            name: secrets
      paths:
      - tkg-templates/providers/infrastructure-aws/v0.6.3/ytt
      - tkg-templates/providers/infrastructure-aws/ytt
      - tkg-templates/providers/ytt
      - tkg-templates/bom
      - tkg-templates/providers/config_default.yaml
      rawOptions: ["--file-mark 'bom**/*:type=text-plain'"]`
```
or possibly?

```yaml
template:
  - ytt:
      ignoreUnknownComments: true
      inline:
        pathsFrom:
        - secretRef:
            name: secrets
      paths:
      - tkg-templates/providers/infrastructure-aws/v0.6.3/ytt
      - tkg-templates/providers/infrastructure-aws/ytt
      - tkg-templates/providers/ytt
      - tkg-templates/bom
      - tkg-templates/providers/config_default.yaml
      file-marks:
         text-plain: 
          - 'bom**/*
```

### PR DESCRIPTION
**Pull request by danielhelfand**
Closes #121 

This pull request adds a fileMarks option to the ytt template section of the App spec to allow users to pass a list of file-marks to `ytt`. 

The format for users will be as follows:

```
template:
  - ytt:
      paths:
      - tkg-templates/providers/infrastructure-aws/v0.6.3/ytt
      - tkg-templates/providers/infrastructure-aws/ytt
      - tkg-templates/providers/ytt
      - tkg-templates/bom
      - tkg-templates/providers/config_default.yaml
      fileMarks:
      - "bom**/*:type=text-plain"
```

### CODE REVIEW

File: pkg/template/ytt.go [hunk 1]
```diff
@@ -129,3 +131,16 @@ func (t *Ytt) addInlinePaths(args []string) ([]string, *memdir.TmpDir, error) {
 
 	return args, inlineDir, nil
 }
+
+func (t *Ytt) addFileMarks(args []string) []string {
+	fileMarks := t.opts.FileMarks
+	if len(fileMarks) == 0 {
+		return args
+	}
+
```
Comment (cppforlife) [inline pkg/template/ytt.go]: dont need this section, since loop will be a noop is there are 0 file marks.

File: test/e2e/template_test.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+	cleanUp := func() {
+		kapp.Run([]string{"delete", "-a", name})
+	}
+
+	cleanUp()
+	defer cleanUp()
+
+	logger.Section("deploy", func() {
+		kapp.RunWithOpts([]string{"deploy", "-f", "-", "-a", name}, RunOpts{StdinReader: strings.NewReader(appYaml)})
```
Comment (cppforlife) [inline test/e2e/template_test.go:62]: this already checks against ReconcileSucceeded. i wonder if its best to check that configmap exists in the cluster.

File: pkg/template/ytt.go [hunk 3]
```diff
@@ -129,3 +131,12 @@ func (t *Ytt) addInlinePaths(args []string) ([]string, *memdir.TmpDir, error) {
 
 	return args, inlineDir, nil
 }
+
+func (t *Ytt) addFileMarks(args []string) []string {
+	fileMarks := t.opts.FileMarks
```
Comment (cppforlife) [inline pkg/template/ytt.go]: this seems like it could be inlined.

**Comment by danielhelfand** [comment]
Making a note that this should also be added to carvel.dev for App spec docs

**Pull request merged**